[1.16] Adds workflow limitations (#4865)
dapr/docs: daprdocs/content/en/reference/components-reference/supported-state-stores/setup-azure-cosmosdb.md

{{% /alert %}}

## Workflow Limitations

> The more complex a workflow is (number of activities, child workflows, etc.), the more state operations it will perform per state store transaction.

Reviewer: ```suggestion
The more complex a workflow is with number of activities, child workflows, etc, the more DB state operations it performs per state store transaction.
```

> All input & output values will also be saved to the workflow history, and will be part of an operation of these transactions.

Reviewer: ```suggestion
All input & output values are saved to the workflow history, and are part of an operation of these transactions.
```

CosmosDB has a [maximum document size of 2MB and maximum transaction size of 100 operations.](https://learn.microsoft.com/en-us/azure/cosmos-db/concepts-limits#per-request-limits).
This means that the workflow history must not exceed this size, meaning that CosmosDB is not suitable for workflows with large input/output values or larger complex workflows.

> 

Reviewer: @JoshVanL - can we provide this info ?
1) A clear description of the error thrown when this limit is reached and how do I handle this either in my code or by looking at the logs
2) Some way that I can provide a simple calculation. For example can we say that typically a workflow will need to stay below X number of activities etc.

{{% /alert %}}

## Workflow Limitations

> The more complex a workflow is with number of activities, child workflows, etc, the more DB state operations it performs per state store transaction.

Reviewer: Suggesting adding here that CosmosDB is not recommended for production when using workflow. WDYT?

All input & output values are saved to the workflow history, and are part of an operation of these transactions.
CosmosDB has a [maximum document size of 2MB and maximum transaction size of 100 operations.](https://learn.microsoft.com/azure/cosmos-db/concepts-limits#per-request-limits).

> Attempting to write to CosmosDB beyond these limits will result in an error code of `413`.

Reviewer: ```suggestion
Attempting to write to CosmosDB beyond these limits results in an error code of `413`.
```

{{% /alert %}}

## Workflow Limitations

{{% alert title="Note" color="primary" %}}

As described below, CosmosDB has limitations that likely make it unsuitable for production environments.

> There is currently no path for migrating Workflow data from CosmosDB to another state store, meaning exceeding these limits in production will result in failed workflows this no workaround.

Reviewer: ```suggestion
There is currently no path for migrating Workflow data from CosmosDB to another state store, meaning exceeding these limits in production will result in failed workflows with no workaround.
```
